intermediate_source/nvfuser_intro_tutorial.py

```diff
@@ -0,0 +1,647 @@
+# `memory_efficient_fusion` pass in FuncTorch, however, is such an
+# optimization pass. To use this pass we have to redefine our
+# function to pull the constants inside (for now it’s easiest to make
+# non-tensor constants literals in the function definition):
+#
+
+
+def primitive_definition_for_memory_efficient_fusion(
+    input1: torch.Tensor,
+    input2: torch.Tensor,
+    weight: torch.Tensor,
+    bias1: torch.Tensor,
+    bias2: torch.Tensor,
+) -> torch.Tensor:
+    bias1_out = input1 + bias1
+    dropout_out = F.dropout(bias1_out, 0.1, training=True)
+    norm_input = dropout_out + input2
+    mean = norm_input.mean(2, keepdim=True)
+    diff = (norm_input - mean)
+    diff_sq = diff * diff
+    var = diff_sq.mean(2, keepdim=True)
+    pre_shift_scale_norm_output = (norm_input - mean) / torch.sqrt(var + 1e-12)
+    norm_output = weight * pre_shift_scale_norm_output + bias2
+    return norm_output
+
+######################################################################
+# Now, instead of passing our function to TorchScript we will pass it
+# to FuncTorch’s optimization pass.
+#
+
+
+# Optimize the model with FuncTorch tracing and the memory efficiency
+# optimization pass
+memory_efficient_primitive_definition = memory_efficient_fusion(
+    primitive_definition_for_memory_efficient_fusion)
+
+# Profile memory efficient primitive definition
+func = functools.partial(
+    memory_efficient_primitive_definition,
+    input1,
+    input2,
+    weight,
+    bias1,
+    bias2)
+profile_workload(func, grad_output, iteration_count=100)
+
+######################################################################
+# This recovers even more speed, but it’s still not as fast as
+# TorchScripts original performance with the composite definition.
+# However, this is still faster than running this new definition
+# without nvFuser, and is still faster than the composite definition
+# without nvFuser.
+#
+# .. figure:: /_static/img/nvfuser_intro/nvfuser_tutorial_5.png
+#
+# .. note:: FuncTorch’s memory efficient pass is still actively in development
+#           and future versions are expected to achieve performance closer
+#           to that of TorchScript with the composite definition.
+
+######################################################################
+# Transformer Block With a Novel Normalization
+# ----------------------------------------------------
+# The ability to quickly execute chains of simple operations is
+# important as not every operation has a composite operation defined
+# in PyTorch. Previously, this meant researchers either had to define
+# an entirely new operation in PyTorch – which takes a lot of time and
+# knowledge of the lower-level PyTorch code as well as parallel
+# programming – or writing the operation in simpler PyTorch ops and
+# settling for poor performance. For example, let's replace LayerNorm
+# in our example with RMSNorm. Even though RMSNorm is a bit simpler
```

Comment:
qq: should we explain RMSNorm itself or add some reference for RMSNorm or is it know enough now?

Reply:
Sounds like a good idea. Would this be the reference? https://arxiv.org/abs/1910.07467?context=cs

Reply:
I think so according to https://github.com/NVIDIA/apex/issues/1271, though, https://doi.org/10.48550/arXiv.1910.07467 might be better